src/libstd/sys/windows/mod.rs

```diff
@@ -72,10 +72,17 @@ pub fn decode_error_kind(errno: i32) -> ErrorKind {
     }
 }
 
-fn to_utf16_os(s: &OsStr) -> Vec<u16> {
-    let mut v: Vec<_> = s.encode_wide().collect();
-    v.push(0);
-    v
+pub fn to_u16s<S: AsRef<OsStr>>(s: &S) -> io::Result<Vec<u16>> {
+    fn inner(s: &OsStr) -> io::Result<Vec<u16>> {
```

Comment:
In general this pattern isn't necessary for internal functions like this, I don't think there's much savings in terms of compile times for the standard library for example (and this definitely won't affect downstream code)